RecoLocalTracker/SiStripClusterizer/plugins/SiStripRawToClusterGPUKernel.cu

```diff
@@ -0,0 +1,551 @@
+            sameDetRight = false;
+          }
+        }  // while loop
+      }    // testIndex < nStrips
+      clusterIndexLeft[i] = indexLeft;
+      clusterSize[i] = indexRight - indexLeft + 1;
+      clusterDetId[i] = det;
+      firstStrip[i] = stripId[indexLeft];
+      trueCluster[i] =
+          (noiseSquared_i * clusterThresholdSquared <= adcSum_i * adcSum_i) and (clusterSize[i] <= clusterSizeLimit);
+    }  // i < nSeedStripsNC
+    if (i == 0) {
+      clust_data_d->nClusters_ = nSeedStripsNC;
+    }
+  }
+
+  __global__ static void checkClusterConditionGPU(StripDataView *sst_data_d,
+                                                  const ConditionsDeviceView *conditions,
+                                                  SiStripClustersCUDADevice::DeviceView *clust_data_d) {
+    const uint16_t *__restrict__ stripId = sst_data_d->stripId;
+    const auto __restrict__ chanlocs = sst_data_d->chanlocs;
+    const uint16_t *__restrict__ channels = sst_data_d->channel;
+    const uint8_t *__restrict__ adc = sst_data_d->adc;
+    const float minGoodCharge = sst_data_d->minGoodCharge;  //1620.0;
+    const auto nSeedStripsNC = clust_data_d->nClusters_;
+    const auto __restrict__ clusterIndexLeft = clust_data_d->clusterIndex_;
+
+    auto __restrict__ clusterSize = clust_data_d->clusterSize_;
+    auto __restrict__ clusterADCs = clust_data_d->clusterADCs_;
+    auto __restrict__ trueCluster = clust_data_d->trueCluster_;
+    auto __restrict__ barycenter = clust_data_d->barycenter_;
+    auto __restrict__ charge = clust_data_d->charge_;
+
+    constexpr uint16_t stripIndexMask = 0x7FFF;
+
+    const int tid = threadIdx.x;
+    const int bid = blockIdx.x;
+    const int nthreads = blockDim.x;
+    const int i = nthreads * bid + tid;
+
+    if (i < nSeedStripsNC) {
+      if (trueCluster[i]) {
+        const int left = clusterIndexLeft[i];
+        const int size = clusterSize[i];
+
+        if (i > 0 && clusterIndexLeft[i - 1] == left) {
+          trueCluster[i] = 0;  // ignore duplicates
+        } else {
+          float adcSum = 0.0f;
+          int sumx = 0;
+          int suma = 0;
+
+          auto j = 0;
+          for (int k = 0; k < size; k++) {
+            const auto index = left + k;
+            const auto chan = channels[index];
+            const auto fed = chanlocs->fedID(chan);
+            const auto channel = chanlocs->fedCh(chan);
+            const auto strip = stripId[index];
+#ifdef GPU_CHECK
+            if (fed == stripgpu::invalidFed) {
+              printf("Invalid fed index %d\n", index);
+            }
+#endif
+            if (strip != stripgpu::invalidStrip) {
+              const float gain_j = conditions->gain(fed, channel, strip);
+
+              uint8_t adc_j = adc[index];
+              const int charge = static_cast<int>(static_cast<float>(adc_j) / gain_j + 0.5f);
+
+              if (adc_j < 254) {
+                adc_j = (charge > 1022 ? 255 : (charge > 253 ? 254 : charge));
+              }
```

Comment:
magic numbers

Reply:
copied from https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/RecoLocalTracker/SiStripClusterizer/src/ThreeThresholdAlgorithm.cc#L114
Does @cms-sw/trk-dpg-l2 have suggestions for constexpr names for these numbers? I can add them.

Reply:
both 254 and 255 ADCs are saturation bins. 254 is the low saturation regime (254 <= raw ADC < 1203), while  255 is the high saturation bin (raw ADC >= 1024). I guess you could call it along these lines.

Reply:
I ended up with
```
             constexpr uint8_t adc_low_saturation = 254;
             constexpr uint8_t adc_high_saturation = 255;
             constexpr int charge_low_saturation = 253;
             constexpr int charge_high_saturation = 1022;
              if (adc_j < adc_low_saturation) {
                adc_j = (charge > charge_high_saturation ? adc_high_saturation : (charge > charge_low_saturation ? adc_low_saturation : charge));
               }

```
makes sense?

Reply:
no objections on my side. In the long run it would be good to make those constants available at a higher level, such that they are used also in the legacy code - but clearly it would be some work to do and outside of this PR scope.